consensus/core/src/core.rs

```diff
@@ -852,6 +952,79 @@ impl Core {
         true
     }
 
+    // Try to decide which of the certified commits will have to be committed next respecting the `limit`. If provided `limit` is zero, it will panic.
+    // The function returns the list of decided leaders and updates in place the remaining certified commits. If empty vector is returned, it means that
+    // there are no certified commits to be committed as `certified_commits` is either empty or all of the certified commits are already committed.
+    #[tracing::instrument(skip_all)]
+    fn try_decide_certified(
+        &mut self,
+        certified_commits: &mut Vec<CertifiedCommit>,
+        limit: usize,
+    ) -> Vec<(DecidedLeader, CertifiedCommit)> {
+        // If GC is disabled then should not run any of this logic.
+        if !self.dag_state.read().gc_enabled() {
+            return Vec::new();
+        }
+
+        assert!(limit > 0, "limit should be greater than 0");
+
+        let last_commit_index = self.dag_state.read().last_commit_index();
+
+        // If there are certified committed leaders, check that the first certified committed leader which is higher than the last decided one has no gaps.
+        while !certified_commits.is_empty() {
+            let certified_commit = certified_commits
+                .first()
+                .expect("Synced commits should not be empty");
+            if certified_commit.index() <= last_commit_index {
+                tracing::debug!(
+                    "Skip commit for index {} as it is already committed with last commit index {}",
+                    certified_commit.index(),
+                    last_commit_index
+                );
+                certified_commits.remove(0);
+            } else {
+                // Make sure that the first commit we find is the next one in line and there is no gap.
+                if certified_commit.index() != last_commit_index + 1 {
```

Comment:
I think this check is pretty critical to the correctness of Core::add_certified_commits(). Personally I would be comfortable having it called first thing in `Core::add_certified_commits()` or included in the function itself. I think try_decide_certified() can assume the CertifiedCommits have been filtered and validated, and can just assert_eq(...). Wdyt?